src/Pulsar.Client.Otel/OtelConsumerInterceptor.fs

```diff
@@ -0,0 +1,81 @@
+
+open System
+open System.Collections.Concurrent
+open System.Collections.Generic
+open System.Diagnostics
+open System.Linq
+open OpenTelemetry
+open OpenTelemetry.Context.Propagation
+open Pulsar.Client.Api
+open Pulsar.Client.Common
+
+type OTelConsumerInterceptor<'T>() =    
+    let Propagator = Propagators.DefaultTextMapPropagator
+    static let  source = "pulsar.consumer"
+    
+    let endActivity(consumer:IConsumer<'T>, messageID:MessageId, ``exception``:Exception, ackType, 
+                            act : ActivitySource, c:ConcurrentDictionary<MessageId,Activity>)=              
+       
+        let gotActFromCache = c.ContainsKey(messageID)
+        let attachOkTagsAndStop (act:Activity) =
+             act.SetTag("acknowledge.type", ackType).
+                                 SetTag("messaging.destination_kind", "topic").
+                                 SetTag("messaging.destination", consumer.Topic).
+                                 SetTag("messaging.message_id", messageID).
+                                 SetTag("messaging.operation", "AfterConsume") |> ignore
+             act.Stop()
+        
+        if gotActFromCache then
+                c.[messageID] |> attachOkTagsAndStop   
+        else
+            let activity = act.StartActivity(consumer.Topic + " consumed",ActivityKind.Consumer)
+            match ``exception`` with
+                    | null ->
+                        activity |> attachOkTagsAndStop   
+                    | _ ->
+                        activity.SetTag("exception.type", ``exception``.Source).
+                                 SetTag("exception.message", ``exception``.Message).
+                                 SetTag("exception.stacktrace", ``exception``.StackTrace) |> ignore
+                        activity.Stop()  
+        ()
+    let getter = Func<IReadOnlyDictionary<string,string>,string,IEnumerable<string>>(fun dict key ->
+                        match dict.TryGetValue(key) with
+                        | true, v -> [v] :> IEnumerable<string>
+                        | false, _ -> Enumerable.Empty<string>()
+                        )
+    let cache = ConcurrentDictionary<MessageId,Activity>()
+    member this.activitySource : ActivitySource = new ActivitySource(source)
+    
+    static member Source
+        with get() = source
+    interface IConsumerInterceptor<'T> with
+        member this.BeforeConsume(consumer, message) =
+            /// Extract the PropagationContext of the upstream parent from the message headers.
+            let mutableDict = message.Properties
+            let contextToInject = Unchecked.defaultof<PropagationContext>   //https://stackoverflow.com/questions/2246206/what-is-the-equivalent-in-f-of-the-c-sharp-default-keyword
+            let parentContext = Propagator.Extract(contextToInject,mutableDict,getter)
+            Baggage.Current <- parentContext.Baggage //baggage is empty for some reason even I parsed metadata from headers
```

Comment:
Why do we need to extract baggage?

Reply:
taken from here https://github.com/open-telemetry/opentelemetry-dotnet/blob/a25741030f05c60c85be102ce7c33f3899290d49/examples/MicroserviceExample/Utils/Messaging/MessageReceiver.cs#L62   In my checks it is always empty though